en/haskell-rust-ffi.md

```diff
@@ -0,0 +1,233 @@
+# Rust FFI in Haskell
+
+This #[[tutorial|tutorial]] will guide you through using [[nix]] to simplify the workflow of incorporating [[rust]] library as a dependency in your [[haskell]] project via [FFI](https://en.wikipedia.org/wiki/Foreign_function_interface). If you are new to [[nix]] and [[flakes]], I recommend starting with the [[nix-tutorial]].
+
+> [!info] Foreign Function Interface (FFI)
+> This isn't solely restricted to Haskell and Rust, it can be used between any two languages that can establish a common ground to communicate, such as C.
+
+The objective of this tutorial is to demonstrate calling a Rust function that returns `Hello, from rust!` from within a Haskell package. Let's begin by setting up the Rust library.
+
+{#init-rust}
+## Initialize Rust Project
+
+Start by initializing a new Rust project using [rust-nix-template](https://github.com/srid/rust-nix-template):
+
+```sh
+git clone https://github.com/srid/rust-nix-template.git
+cd rust-nix-template
+```
+
+Now, let's run the project:
+
+```sh
+nix develop
+just run
+```
+
+{#rust-lib}
+## Create a Rust Library
+
+The template we've initialized is a binary project, but we need a library project. The library should export a function callable from Haskell. For simplicity, let's export a function named `hello` that returns a `C-style string`. To do so, create a new file named `src/lib.rs` with the following contents:
+
+[[haskell-rust-ffi/lib.rs]]
+![[haskell-rust-ffi/lib.rs]]
+
+> [!info] Calling Rust code from C
+> You can learn more about it [here](https://doc.rust-lang.org/nomicon/ffi.html#calling-rust-code-from-c).
+
+Now, the library builds, but we need the dynamic library files required for FFI. To achieve this, let's add a `crate-type` to the `Cargo.toml`:
+
+```toml
+[lib]
+crate-type = ["cdylib"]
+```
+
+After running `cargo build`, you should find a `librust_nix_template.dylib`[^hyphens-disallowed] (if you are on macOS) or `librust_nix_template.so` (if you are on Linux) in the `target/debug` directory.
+
+[^hyphens-disallowed]: Note that the hyphens are disallowed in the library name; hence it's named `librust_nix_template.dylib`. Explicitly setting the name of the library with hyphens will fail while parsing the manifest with: `library target names cannot contain hyphens: rust-nix-template`
+
+{#init-haskell}
+## Initialize Haskell Project
+
+Fetch `cabal-install` and `ghc` from the `nixpkgs` input of current flake and initialize a new Haskell project[^why-proj-nixpkgs]:
+
+[^why-proj-nixpkgs]: This approach ensures reproducibility as `cabal init` uses the version of GHC to initialize the `base` package constraints.
+
+```sh
+nixpkgs_url="github:nixos/nixpkgs/$(nix flake metadata --json | nix run nixpkgs#jq -- '.locks.nodes.nixpkgs.locked.rev' -r)" && \
+nix shell $nixpkgs_url#cabal-install $nixpkgs_url#ghc -c \
+cabal init -n --exe -m --simple hello-haskell
```

Comment:
This will break the further steps in the tutorial, as we are assuming that the `.cabal` file generated by `cabal-install` in `flake:nixpkgs` will work with the `nixpkgs` input of the flake in the tutorial.

This was the initial command, but when I was following along the tutorial it fails, as constraints on `base` depends on `GHC` version, which can be different based on the nixpkgs commit. I doubt there is a way to `cabal init` without any constraints on the `base` package.

Reply:
Pass `-d base` option?

```
❯ nix shell nixpkgs#ghc nixpkgs#cabal-install -c cabal -- init -n --exe -m --simple -d base --overwrite
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.
[Log] Using cabal specification: 3.0
[Warning] unknown license type, you must put a copy in LICENSE yourself.
[Log] CHANGELOG.md already exists. Backing up old version in CHANGELOG.md.save0
[Log] Overwriting file CHANGELOG.md...
[Log] app already exists. Backing up old version in app.save1
[Log] Overwriting directory ./app...
[Log] Creating fresh file app/Main.hs...
[Log] hello-haskell.cabal already exists. Backing up old version in hello-haskell.cabal.save0
[Log] Overwriting file hello-haskell.cabal...
[Warning] No synopsis given. You should edit the .cabal file and add one.
[Info] You may want to edit the .cabal file and add a Description field.


srid on appreciate.intern /tmp/hello-haskell via λ
❯ rg base *.cabal
14:    build-depends:    base
```

Reply:
Thanks!

Reply:
https://github.com/nixos-asia/website/pull/24/commits/b28fabe260c971bfed4d7dea3aa84c8106321ffc